File: 2021/haskell/Day0101.hs
module Main where

import Data.Char
import Text.Regex.TDFA
import Text.Regex.TDFA.Text

nthEven :: Int -> Int
nthEven n = 2 * n - 2

validateEmail :: String -> Bool
validateEmail = flip (=~) ".+@.+\\..+"

atbash :: String -> String
atbash = map atbash_letter
  where
    atbash_letter l
      | isLower l = chr $ ord 'z' - ord l + ord 'a'
      | isUpper l = chr $ ord 'Z' - ord l + ord 'A'
      | otherwise = l

main :: IO ()
main = do
  putStrLn ("nth_even #1: " ++ (show $ nthEven 1))
  putStrLn ("nth_even #2: " ++ (show $ nthEven 2))
  putStrLn ("nth_even #3: " ++ (show $ nthEven 100))
  putStrLn "----------"
  putStrLn ("validateEmail #1: " ++ (show $ validateEmail "@gmail.com"))
  putStrLn ("validateEmail #2: " ++ (show $ validateEmail "hello.gmail@com"))
  putStrLn ("validateEmail #3: " ++ (show $ validateEmail "gmail"))
  putStrLn ("validateEmail #4: " ++ (show $ validateEmail "hello@gmail"))
  putStrLn ("validateEmail #5: " ++ (show $ validateEmail "hello@edabit.com"))
  putStrLn "----------"
  putStrLn ("atbash #1: " ++ atbash "apple")
  putStrLn ("atbash #2: " ++ atbash "Hello world!")
  putStrLn ("atbash #3: " ++ atbash "Christmas is the 25th of December")

File: 2021/clojure/day0101.clj
(comment
  "Author: Loading BG#7962"
  )

(defn nth-even [n]
  (- (* 2 n) 2))

(defn validate-email [email]
  (if (re-matches #".+@.+\..+" email) true false))

(defn- atbash-letter [l]
  (cond
    (Character/isUpperCase l) (char (+ (- (int \Z) (int l)) (int \A)))
    (Character/isLowerCase l) (char (+ (- (int \z) (int l)) (int \a)))
    :else l))
(defn atbash [s]
  (apply str (map atbash-letter s)))

(println "nth_even #1:" (nth-even 1))
(println "nth_even #2:" (nth-even 2))
(println "nth_even #3:" (nth-even 100))
(println "------------")
(println "validateEmail #1:" (validate-email "@gmail.com"))
(println "validateEmail #2:" (validate-email "hello.gmail@com"))
(println "validateEmail #3:" (validate-email "gmail"))
(println "validateEmail #4:" (validate-email "hello@gmail"))
(println "validateEmail #5:" (validate-email "hello@edabit.com"))
(println "-----------------")
(println "atbash #1:" (atbash "apple"))
(println "atbash #2:" (atbash "Hello world!"))
(println "atbash #3:" (atbash "Christmas is the 25th of December"))
